File: tests/abi/note_expression_layout.cpp
#include "pluginterfaces/vst/ivstnoteexpression.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Vst::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("NoteExpressionTypeIDs.kVolumeTypeID %u\n", Steinberg::Vst::kVolumeTypeID);
	std::printf ("NoteExpressionTypeIDs.kPanTypeID %u\n", Steinberg::Vst::kPanTypeID);
	std::printf ("NoteExpressionTypeIDs.kTuningTypeID %u\n", Steinberg::Vst::kTuningTypeID);
	std::printf ("NoteExpressionTypeIDs.kVibratoTypeID %u\n", Steinberg::Vst::kVibratoTypeID);
	std::printf ("NoteExpressionTypeIDs.kExpressionTypeID %u\n", Steinberg::Vst::kExpressionTypeID);
	std::printf ("NoteExpressionTypeIDs.kBrightnessTypeID %u\n", Steinberg::Vst::kBrightnessTypeID);
	std::printf ("NoteExpressionTypeIDs.kTextTypeID %u\n", Steinberg::Vst::kTextTypeID);
	std::printf ("NoteExpressionTypeIDs.kPhonemeTypeID %u\n", Steinberg::Vst::kPhonemeTypeID);
	std::printf ("NoteExpressionTypeIDs.kCustomStart %u\n", Steinberg::Vst::kCustomStart);
	std::printf ("NoteExpressionTypeIDs.kCustomEnd %u\n", Steinberg::Vst::kCustomEnd);
	std::printf ("NoteExpressionTypeIDs.kInvalidTypeID %u\n", Steinberg::Vst::kInvalidTypeID);
	std::printf ("NoteExpressionTypeInfo.kIsBipolar %d\n", Steinberg::Vst::NoteExpressionTypeInfo::kIsBipolar);
	std::printf ("NoteExpressionTypeInfo.kIsOneShot %d\n", Steinberg::Vst::NoteExpressionTypeInfo::kIsOneShot);
	std::printf ("NoteExpressionTypeInfo.kIsAbsolute %d\n", Steinberg::Vst::NoteExpressionTypeInfo::kIsAbsolute);
	std::printf ("NoteExpressionTypeInfo.kAssociatedParameterIDValid %d\n", Steinberg::Vst::NoteExpressionTypeInfo::kAssociatedParameterIDValid);
	std::printf ("KeyswitchTypeIDs.kNoteOnKeyswitchTypeID %u\n", Steinberg::Vst::kNoteOnKeyswitchTypeID);
	std::printf ("KeyswitchTypeIDs.kOnTheFlyKeyswitchTypeID %u\n", Steinberg::Vst::kOnTheFlyKeyswitchTypeID);
	std::printf ("KeyswitchTypeIDs.kOnReleaseKeyswitchTypeID %u\n", Steinberg::Vst::kOnReleaseKeyswitchTypeID);
	std::printf ("KeyswitchTypeIDs.kKeyRangeTypeID %u\n", Steinberg::Vst::kKeyRangeTypeID);

	PRINT_TYPE (NoteExpressionValueEvent);
	PRINT_OFFSET (NoteExpressionValueEvent, typeId);
	PRINT_OFFSET (NoteExpressionValueEvent, noteId);
	PRINT_OFFSET (NoteExpressionValueEvent, value);

	PRINT_TYPE (NoteExpressionIntValueEvent);
	PRINT_OFFSET (NoteExpressionIntValueEvent, typeId);
	PRINT_OFFSET (NoteExpressionIntValueEvent, noteId);
	PRINT_OFFSET (NoteExpressionIntValueEvent, value);

	PRINT_TYPE (NoteExpressionTextEvent);
	PRINT_OFFSET (NoteExpressionTextEvent, typeId);
	PRINT_OFFSET (NoteExpressionTextEvent, noteId);
	PRINT_OFFSET (NoteExpressionTextEvent, textLen);
	PRINT_OFFSET (NoteExpressionTextEvent, text);

	PRINT_TYPE (NoteExpressionValueDescription);
	PRINT_OFFSET (NoteExpressionValueDescription, defaultValue);
	PRINT_OFFSET (NoteExpressionValueDescription, minimum);
	PRINT_OFFSET (NoteExpressionValueDescription, maximum);
	PRINT_OFFSET (NoteExpressionValueDescription, stepCount);

	PRINT_TYPE (NoteExpressionTypeInfo);
	PRINT_OFFSET (NoteExpressionTypeInfo, typeId);
	PRINT_OFFSET (NoteExpressionTypeInfo, title);
	PRINT_OFFSET (NoteExpressionTypeInfo, shortTitle);
	PRINT_OFFSET (NoteExpressionTypeInfo, units);
	PRINT_OFFSET (NoteExpressionTypeInfo, unitId);
	PRINT_OFFSET (NoteExpressionTypeInfo, valueDesc);
	PRINT_OFFSET (NoteExpressionTypeInfo, associatedParameterId);
	PRINT_OFFSET (NoteExpressionTypeInfo, flags);

	PRINT_TYPE (KeyswitchInfo);
	PRINT_OFFSET (KeyswitchInfo, typeId);
	PRINT_OFFSET (KeyswitchInfo, title);
	PRINT_OFFSET (KeyswitchInfo, shortTitle);
	PRINT_OFFSET (KeyswitchInfo, keyswitchMin);
	PRINT_OFFSET (KeyswitchInfo, keyswitchMax);
	PRINT_OFFSET (KeyswitchInfo, keyRemapped);
	PRINT_OFFSET (KeyswitchInfo, unitId);
	PRINT_OFFSET (KeyswitchInfo, flags);

	print_iid ("INoteExpressionController", Steinberg::Vst::INoteExpressionController_iid);
	print_iid ("IKeyswitchController", Steinberg::Vst::IKeyswitchController_iid);
	return 0;
}

File: tools/note_expression_layout.zig
const std = @import("std");
const note_expression = @import("vst3-zig").pluginterfaces.vst.ivstnoteexpression;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("NoteExpressionTypeIDs.kVolumeTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kVolumeTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kPanTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kPanTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kTuningTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kTuningTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kVibratoTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kVibratoTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kExpressionTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kExpressionTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kBrightnessTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kBrightnessTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kTextTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kTextTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kPhonemeTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kPhonemeTypeID)});
    try stdout.print("NoteExpressionTypeIDs.kCustomStart {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kCustomStart)});
    try stdout.print("NoteExpressionTypeIDs.kCustomEnd {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kCustomEnd)});
    try stdout.print("NoteExpressionTypeIDs.kInvalidTypeID {}\n", .{@intFromEnum(note_expression.NoteExpressionTypeIDs.kInvalidTypeID)});
    try stdout.print("NoteExpressionTypeInfo.kIsBipolar {}\n", .{note_expression.NoteExpressionTypeInfo.NoteExpressionTypeFlags.kIsBipolar});
    try stdout.print("NoteExpressionTypeInfo.kIsOneShot {}\n", .{note_expression.NoteExpressionTypeInfo.NoteExpressionTypeFlags.kIsOneShot});
    try stdout.print("NoteExpressionTypeInfo.kIsAbsolute {}\n", .{note_expression.NoteExpressionTypeInfo.NoteExpressionTypeFlags.kIsAbsolute});
    try stdout.print("NoteExpressionTypeInfo.kAssociatedParameterIDValid {}\n", .{note_expression.NoteExpressionTypeInfo.NoteExpressionTypeFlags.kAssociatedParameterIDValid});
    try stdout.print("KeyswitchTypeIDs.kNoteOnKeyswitchTypeID {}\n", .{@intFromEnum(note_expression.KeyswitchTypeIDs.kNoteOnKeyswitchTypeID)});
    try stdout.print("KeyswitchTypeIDs.kOnTheFlyKeyswitchTypeID {}\n", .{@intFromEnum(note_expression.KeyswitchTypeIDs.kOnTheFlyKeyswitchTypeID)});
    try stdout.print("KeyswitchTypeIDs.kOnReleaseKeyswitchTypeID {}\n", .{@intFromEnum(note_expression.KeyswitchTypeIDs.kOnReleaseKeyswitchTypeID)});
    try stdout.print("KeyswitchTypeIDs.kKeyRangeTypeID {}\n", .{@intFromEnum(note_expression.KeyswitchTypeIDs.kKeyRangeTypeID)});

    try printType(stdout, "NoteExpressionValueEvent", note_expression.NoteExpressionValueEvent);
    try printOffset(stdout, "NoteExpressionValueEvent", "typeId", note_expression.NoteExpressionValueEvent, "typeId");
    try printOffset(stdout, "NoteExpressionValueEvent", "noteId", note_expression.NoteExpressionValueEvent, "noteId");
    try printOffset(stdout, "NoteExpressionValueEvent", "value", note_expression.NoteExpressionValueEvent, "value");

    try printType(stdout, "NoteExpressionIntValueEvent", note_expression.NoteExpressionIntValueEvent);
    try printOffset(stdout, "NoteExpressionIntValueEvent", "typeId", note_expression.NoteExpressionIntValueEvent, "typeId");
    try printOffset(stdout, "NoteExpressionIntValueEvent", "noteId", note_expression.NoteExpressionIntValueEvent, "noteId");
    try printOffset(stdout, "NoteExpressionIntValueEvent", "value", note_expression.NoteExpressionIntValueEvent, "value");

    try printType(stdout, "NoteExpressionTextEvent", note_expression.NoteExpressionTextEvent);
    try printOffset(stdout, "NoteExpressionTextEvent", "typeId", note_expression.NoteExpressionTextEvent, "typeId");
    try printOffset(stdout, "NoteExpressionTextEvent", "noteId", note_expression.NoteExpressionTextEvent, "noteId");
    try printOffset(stdout, "NoteExpressionTextEvent", "textLen", note_expression.NoteExpressionTextEvent, "textLen");
    try printOffset(stdout, "NoteExpressionTextEvent", "text", note_expression.NoteExpressionTextEvent, "text");

    try printType(stdout, "NoteExpressionValueDescription", note_expression.NoteExpressionValueDescription);
    try printOffset(stdout, "NoteExpressionValueDescription", "defaultValue", note_expression.NoteExpressionValueDescription, "defaultValue");
    try printOffset(stdout, "NoteExpressionValueDescription", "minimum", note_expression.NoteExpressionValueDescription, "minimum");
    try printOffset(stdout, "NoteExpressionValueDescription", "maximum", note_expression.NoteExpressionValueDescription, "maximum");
    try printOffset(stdout, "NoteExpressionValueDescription", "stepCount", note_expression.NoteExpressionValueDescription, "stepCount");

    try printType(stdout, "NoteExpressionTypeInfo", note_expression.NoteExpressionTypeInfo);
    try printOffset(stdout, "NoteExpressionTypeInfo", "typeId", note_expression.NoteExpressionTypeInfo, "typeId");
    try printOffset(stdout, "NoteExpressionTypeInfo", "title", note_expression.NoteExpressionTypeInfo, "title");
    try printOffset(stdout, "NoteExpressionTypeInfo", "shortTitle", note_expression.NoteExpressionTypeInfo, "shortTitle");
    try printOffset(stdout, "NoteExpressionTypeInfo", "units", note_expression.NoteExpressionTypeInfo, "units");
    try printOffset(stdout, "NoteExpressionTypeInfo", "unitId", note_expression.NoteExpressionTypeInfo, "unitId");
    try printOffset(stdout, "NoteExpressionTypeInfo", "valueDesc", note_expression.NoteExpressionTypeInfo, "valueDesc");
    try printOffset(stdout, "NoteExpressionTypeInfo", "associatedParameterId", note_expression.NoteExpressionTypeInfo, "associatedParameterId");
    try printOffset(stdout, "NoteExpressionTypeInfo", "flags", note_expression.NoteExpressionTypeInfo, "flags");

    try printType(stdout, "KeyswitchInfo", note_expression.KeyswitchInfo);
    try printOffset(stdout, "KeyswitchInfo", "typeId", note_expression.KeyswitchInfo, "typeId");
    try printOffset(stdout, "KeyswitchInfo", "title", note_expression.KeyswitchInfo, "title");
    try printOffset(stdout, "KeyswitchInfo", "shortTitle", note_expression.KeyswitchInfo, "shortTitle");
    try printOffset(stdout, "KeyswitchInfo", "keyswitchMin", note_expression.KeyswitchInfo, "keyswitchMin");
    try printOffset(stdout, "KeyswitchInfo", "keyswitchMax", note_expression.KeyswitchInfo, "keyswitchMax");
    try printOffset(stdout, "KeyswitchInfo", "keyRemapped", note_expression.KeyswitchInfo, "keyRemapped");
    try printOffset(stdout, "KeyswitchInfo", "unitId", note_expression.KeyswitchInfo, "unitId");
    try printOffset(stdout, "KeyswitchInfo", "flags", note_expression.KeyswitchInfo, "flags");

    try printTuid(stdout, "INoteExpressionController", note_expression.inote_expression_controller_iid);
    try printTuid(stdout, "IKeyswitchController", note_expression.ikeyswitch_controller_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
